rpc/src/eth.rs

```diff
@@ -228,6 +241,86 @@ impl<B, C, P, CT, BE, H: ExHashT> EthApi<B, C, P, CT, BE, H> where
 		}
 		Ok(None)
 	}
+
+	fn current_block(&self, id: &BlockId<B>) -> Option<ethereum::Block> {
+		self.query_storage::<ethereum::Block>(
+			id,
+			&StorageKey(
+				storage_prefix_build(b"Ethereum", b"CurrentBlock")
+			)
+		)
+	}
+
+	fn current_statuses(&self, id: &BlockId<B>) -> Option<Vec<TransactionStatus>> {
+		self.query_storage::<Vec<TransactionStatus>>(
+			id,
+			&StorageKey(
+				storage_prefix_build(b"Ethereum", b"CurrentTransactionStatuses")
+			)
+		)
+	}
+
+	fn current_receipts(&self, id: &BlockId<B>) -> Option<Vec<ethereum::Receipt>> {
+		self.query_storage::<Vec<ethereum::Receipt>>(
+			id,
+			&StorageKey(
+				storage_prefix_build(b"Ethereum", b"CurrentReceipts")
+			)
+		)
+	}
+
+	fn account_codes(&self, id: &BlockId<B>, address: H160) -> Option<Vec<u8>> {
+		let mut key: Vec<u8> = storage_prefix_build(b"EVM", b"AccountCodes");
+		key.extend(blake2_128_extend(address.as_bytes()));
+		self.query_storage::<Vec<u8>>(
+			id,
+			&StorageKey(key)
+		)
+	}
+
+	fn account_storages(&self, id: &BlockId<B>, address: H160, index: U256) -> Option<H256> {
+		let tmp: &mut [u8; 32] = &mut [0; 32];
+		index.to_little_endian(tmp);
+
+		let mut key: Vec<u8> = storage_prefix_build(b"EVM", b"AccountStorages");
+		key.extend(blake2_128_extend(address.as_bytes()));
+		key.extend(blake2_128_extend(tmp));
+
+		self.query_storage::<H256>(
+			id,
+			&StorageKey(key)
+		)
+	}
+
+	fn chain_id(&self, id: &BlockId<B>) -> Option<u64> {
+		self.query_storage::<u64>(
+			id,
+			&StorageKey(
+				storage_prefix_build(b"EVM", b"ChainId")
+			)
+		)
+	}
+
+	fn gas_price(&self, id: &BlockId<B>) -> Option<U256> {
+		self.query_storage::<U256>(
+			id,
+			&StorageKey(
+				storage_prefix_build(b"EVM", b"GasPrice")
+			)
+		)
+	}
+
+	fn query_storage<T: Decode>(&self, id: &BlockId<B>, key: &StorageKey) -> Option<T> {
+		if let Ok(Some(data)) = self.client.storage(
+			id,
+			key
+		) {
+			if let Ok(result) = Decode::decode(&mut &data.0[..]) {
+				return Some(result);
+			}
+		}
+		None
+	}
```

Comment:
Overall I love this generically typed function.

I think it could be improved by returning a Result rather than an Option. This could fail in two different ways. First, looking up the storage in the block could fail, and second, decoding could fail. So it is nice to pass that information on to the caller.
```rust
fn query_storage<T: Decode>(&self, id: &BlockId<B>, key: &StorageKey) -> Result<T, StorageError>;
```

Of course the RPC will eventually just have to return an option, but that should be handled in each rpc method using [`Result.ok`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok)